README.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      README.md
--
--  BRIEF
--      Important information regarding this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--
------------------------------------------------------------------------------->

# untab

## Summary

A simple lexer which removes tab characters from a given file.

## License

This project's license is **GPL 2** (as of June 1991).  The whole license text
can be found in `LICENSE` in the main directory of this repository.  The brief
version is as follows:

> Copyright (C) 2022 Kevin Matthes
>
> This program is free software; you can redistribute it and/or modify
> it under the terms of the GNU General Public License as published by
> the Free Software Foundation; either version 2 of the License, or
> (at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
>
> You should have received a copy of the GNU General Public License along
> with this program; if not, write to the Free Software Foundation, Inc.,
> 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

When compiling a printable version of this documentation using Pandoc, the full
license will be attached automatically to the resulting document.  This can be
invoked by calling `repository-manual.m`.

## Software Requirements

| Requirement       | Type          | Role                                  |
|:------------------|:-------------:|:--------------------------------------|
| `docs-snippets`   | repository    | documentation constants               |
| Flex              | application   | Lex compiler                          |
| GCC               | application   | C compiler                            |
| GNU Octave        | application   | execution of the provided scripts     |
| Pandoc            | application   | compilation of repository manual      |
| `texlive-full`    | package       | compilation of repository manual      |

The compilation of such an **optional** repository manual can be invoked by just
calling one of the following lines in a terminal.

```
octave repository-manual.m
octave-cli repository-manual.m
```

Both will redirect to the same application, GNU Octave, which will then create
the manual for this repository and attach the entire license to it.  The
resulting file, `repository.pdf`, will be saved in the main directory of this
repository.

## Description

This repository provides a simple lexer in order to replace tab characters from
input source files with space characters.  It can be compiled by invoking the
GNU Octave script `flex-gcc.m` in a terminal:

```
octave flex-gcc.m
octave-cli flex-gcc.m
```

This lexer will replace any tab characters in the input stream with space
characters.  In order to define the count of spaces to replace a single tab
with, the command line arguments will be used to determine the intended value.

The intended tab width can be any integer from zero to nine, including zero and
nine.  The desired width needs be to entered as a decimal integer with `-` as a
prefix.  For instance, in order to remove all tab characters, `-0`, for tab
width zero, needs to be passed.  If no argument is passed to this lexer, 4 will
be assumed, by default.  Only one tab width setting is permitted.

This lexer offers both an interactive mode as well as a batch one.

The former is invoked automatically if no file name is passed.  Anything which
does not carry `-` as a prefix is assumed to be a file name.  In the interactive
mode, the lexer reads from `stdin` and writes to `stdout`.  This is useful in
order to pipe data around in a terminal session.

The batch mode is invoked automatically by passing a file name to the lexer.
Anything which does not carry `-` as a prefix is assumed to be a file name.  The
lexer will read from the given file, write the intermediate output to an
auxillary file named `untab.l.aux` and rewrite the original file at the end.
The auxillary file is tidied up automatically.

In case of errors, the lexer will automatically abort the execution and will
close all open streams and free all allocated memory regions, respectively.
Corresponding error messages will be written to `stderr` and the return value of
the lexer will be set according to `sysexits.h`.

Possible errors are:
* mandatory memory allocations fail
* required files cannot be accessed and / or created
* unknown command line options

The meaning of the return values are as follows:

| Code   | Meaning                               |
|:------:|:--------------------------------------|
| 0      | no problem occured                    |
| 64     | unknown option / too many arguments   |
| 69     | required file not accessable          |
| 73     | required file not creatable           |
| 74     | too few memory                        |

The script `install-app.m` will install the application in `~/.local/bin/`, at
option.

```
octave install-app.m
octave-cli install-app.m
```

## Options

### `-0`

```
untab -0
```

Set the tab width to zero.  The default value is four.

Only one setting for the tab width is allowed.

### `-1`

```
untab -1
```

Set the tab width to one.  The default value is four.

Only one setting for the tab width is allowed.

### `-2`

```
untab -2
```

Set the tab width to two.  The default value is four.

Only one setting for the tab width is allowed.

### `-3`

```
untab -3
```

Set the tab width to three.  The default value is four.

Only one setting for the tab width is allowed.

### `-4`

```
untab -4
```

Set the tab width to four.  This is the default setting.

Only one setting for the tab width is allowed.

### `-5`

```
untab -5
```

Set the tab width to five.  The default value is four.

Only one setting for the tab width is allowed.

### `-6`

```
untab -6
```

Set the tab width to six.  The default value is four.

Only one setting for the tab width is allowed.

### `-7`

```
untab -7
```

Set the tab width to seven.  The default value is four.

Only one setting for the tab width is allowed.

### `-8`

```
untab -8
```

Set the tab width to eight.  The default value is four.

Only one setting for the tab width is allowed.

### `-9`

```
untab -9
```

Set the tab width to nine.  The default value is four.

Only one setting for the tab width is allowed.

### `-l`

```
untab -l
```

Show the license text and quit.  All further options will be ignored.

### `-v`

```
untab -v
```

Show the version and quit.  All further options will be ignored.

<!----------------------------------------------------------------------------->


CHANGELOG.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      CHANGELOG.md
--
--  BRIEF
--      The development history of this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--      See `README.md' for project details.
--
------------------------------------------------------------------------------->

# Changelog

## [Unreleased]

## [1.5] -- 2022-04-10

### Fixed

* typos in docstrings
* typos in README description

### Removed

* header `ctypes.h` as include

## [1.4] -- 2022-04-06

### Added

* create `define.dox`
* create `GPL_2_0.h`
* description for all options
* function `creation_failed`
* function `license`
* function `version`
* option for license information
* option for verion information

### Changed

* adjust docstrings
* adjust option parsing
* adjust README description
* call `creation_failed` instead of `unknown_file` when file creations fail
* outsource all macro docstrings to `define.dox`

### Deprecated

* header `ctypes.h` as include

### Fixed

* accept only the first tab width setting
* typos in changelog items
* update submodule `routines`

## [1.3] -- 2022-04-05

### Changed

* adjust docstrings
* adjust README description

### Fixed

* unused functions `input` and `yyunput`

## [1.2] -- 2022-04-04

### Added

* basic option parsing
* forward declaration of `main` in main lexfile
* function `eval`
* function `finish`
* function `memory_insufficient`
* function `prepare`
* function `process`
* function `spaces`
* function `unknown_file`
* function `unknown_option`
* function `unresolvable_argument`
* include `ctypes.h`
* include `sysexits.h`
* optimisation of given files -- given as command line arguments

### Changed

* data type of `tablength` from `unsigned int` to `int`
* outsource argument evaluation to function `eval`
* outsource lexer call and destruction to function `process`
* outsource space character insertion to function `spaces`

### Fixed

* too much `const`

### Removed

* passing integers to the application for setting the tab width

## [1.1] -- 2022-04-02

### Deprecated

* passing integers to the application for setting the tab width

### Fixed

* missing changelog item
* obsolete banners in build routine
* typo in changelog item
* typos in docstrings

## [1.0] -- 2022-04-02

### Added

* build instruction:  installation
* create `install-app.m`

### Fixed

* fix memory leaks of lexer

## [0.4] -- 2022-04-02

### Added

* docstrings for Doxygen
* minimal project description in README
* software requirement:  Flex

### Changed

* data type of `tablength` from `int` to `unsigned int`

### Fixed

* changelog as Appendix B in repository manual
* missing changelog item

### Removed

* software requirement:  Doxygen
* software requirement:  `make`

## [0.3] -- 2022-04-01

### Added

* build instruction:  entire compilation routine
* create `flex-gcc.m`

### Deprecated

* software requirement:  Doxygen
* software requirement:  `make`

## [0.2] -- 2022-04-01

### Added

* create `untab.l`

### Fixed

* ignore `lex.*.c`
* ignore `untab` (compilation target)

## [0.1] -- 2022-04-01

### Added

* add GPL 2.0 license
* build instruction:  repository manual
* create `.gitignore`
* create list of authors
* create project meta data for repository manual
* create repository README
* create this changelog
* create this repository
* submodule `docs-snippets`
* submodule `routines`

<!----------------------------------------------------------------------------->


GPL_2_0.h
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The brief license text of the GPL 2.0 license.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        GPL_2_0.h
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This header defines the brief license text of the GPL 2.0 license as a macro.
 */

/******************************************************************************/

/*
 * Security settings.
 */

#pragma once
#ifndef __GPL_2_0_H__
#define __GPL_2_0_H__



/**
 * \brief   The GPL 2.0 license (brief version).
 *
 * This macro represents the brief version of the GPL 2.0 license text.
 */

#define GPL_2_0 \
    "This program is free software; you can redistribute it and/or modify\n"   \
    "it under the terms of the GNU General Public License as published by\n"   \
    "the Free Software Foundation; either version 2 of the License, or\n"      \
    "(at your option) any later version.\n"                                    \
    "\n"                                                                       \
    "This program is distributed in the hope that it will be useful,\n"        \
    "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"         \
    "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"          \
    "GNU General Public License for more details.\n"                           \
    "\n"                                                                       \
    "You should have received a copy of the GNU General Public License along\n"\
    "with this program; if not, write to the Free Software Foundation, Inc.,\n"\
    "51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA."



/*
 * End of header.
 */

// Leaving the header.
#endif  // ! __GPL_2_0_H__

/******************************************************************************/


flex-gcc.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      flex-gcc.m
%%
%%  BRIEF
%%      Create a new lexer using Flex and GCC.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
compiler.flags  = ' -Wall -Werror -Wextra -Wpedantic ';
compiler.flags  = [compiler.flags ' -DYY_NO_INPUT '];
compiler.flags  = [compiler.flags ' -DYY_NO_UNPUT '];
compiler.in     = '*.c';
compiler.link   = ' -lfl ';
compiler.out    = 'untab';
compiler.self   = ' gcc ';
compiler.call   = [compiler.self ' ' compiler.flags ' ' compiler.in];
compiler.call   = [compiler.call ' ' compiler.link ' -o ' compiler.out];

scangen.in      = 'untab.l';
scangen.out     = 'lex.yy.c';
scangen.self    = ' flex ';
scangen.call    = [scangen.self ' ' scangen.in];



% Files.
files.self  = 'flex-gcc.m';



% Miscellaneous.
banner  = ['[ ' files.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Call Flex.
disp ([banner 'Compile scanner definition with Flex ...']);

disp (scangen.call);
system (scangen.call);

disp ([banner 'Done.']);



% Call GCC.
disp ([banner 'Compile C source code with GCC ...']);

disp (compiler.call);
system (compiler.call);

disp ([banner 'Done.']);



% Clean build artifacts.
fprintf ([banner 'Remove build artifacts ... ']);

if length (glob (scangen.out));
    delete (scangen.out);
end;

disp ('Done.');



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


flex-gcc-compile.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      flex-gcc.m
%%
%%  BRIEF
%%      Create a new lexer using Flex and GCC.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
compiler.self   = ' gcc ';
compiler.flags  = ' -Wall -Werror -Wextra -Wno-unused-function -Wpedantic ';
compiler.link   = ' -lfl ';
compiler.call   = [compiler.self compiler.flags];

scangen.self    = ' flex ';



% Files.
files.self  = 'flex-gcc.m'

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


repository-manual.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      repository-manual.m
%%
%%  BRIEF
%%      Create a LaTeX documentation for this repository using Pandoc.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Settings.
manual.language = 'British';
manual.paper    = 'a4-3x3-duplex';
manual.style    = 'scrartcl';



% Software.
software.compiler.self  = ' pandoc ';
software.compiler.flags = ' -N ';
software.compiler.call  = [software.compiler.self software.compiler.flags];



% Directories.
directories.md      = './docs-snippets/markdown/';
directories.yaml    = './docs-snippets/yaml/';



% Files.
files.code.begin    = [directories.md 'begin-code.md '];
files.code.end      = [directories.md 'end-code.md '];

files.license       = [files.code.begin ' ./LICENSE ' files.code.end];
files.license       = [directories.md 'heading-license.md ' files.license];

files.newpage       = [directories.md 'newpage.md '];

files.self          = ' repository-manual.m ';

files.source        = ' ./authors.yaml ./project.yaml ';
files.source        = [files.source directories.yaml manual.language '.yaml '];
files.source        = [files.source directories.yaml manual.paper '.yaml '];
files.source        = [files.source directories.yaml manual.style '.yaml '];
files.source        = [files.source files.newpage];
files.source        = [files.source ' ./README.md '];
files.source        = [files.source files.newpage];
files.source        = [files.source ' ./CHANGELOG.md '];
files.source        = [files.source files.license];
files.source        = [files.source files.newpage];

files.target        = ' repository.pdf ';



% Control flow.
banner  = ['[' files.self '] '];



% Call adjustment.
software.compiler.call  = [software.compiler.call files.source];
software.compiler.call  = [software.compiler.call ' -o ' files.target];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Call Pandoc.
disp ([banner 'Compile Pandoc documentation ...']);

disp (software.compiler.call);
system (software.compiler.call);

disp ([banner 'Done.']);



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


install-app.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      install-app.m
%%
%%  BRIEF
%%      Install a compilation of the provided application.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
octave.in   = 'flex-gcc.m';
octave.out  = './untab';
octave.self = ' octave ';
octave.call = [octave.self ' ' octave.in];

% Directories.
directories.dep1    = '~/.local/';
directories.target  = '~/.local/bin/';

% Files.
files.self  = 'install-app.m';

% Miscellaneous.
banner  = ['[ ' files.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Compile application.
disp ([banner 'Compile the application to install ...']);
system (octave.call);
disp ([banner 'Done.']);



% Create directories, if required.
fprintf ([banner 'Check for ' directories.dep1 ' ... ']);
if length (glob (directories.dep1));
    disp ('Done.');
else;
    mkdir (directories.dep1);
    disp ('Created.');
end;

fprintf ([banner 'Check for ' directories.target ' ... ']);
if length (glob (directories.target));
    disp ('Done.');
else;
    mkdir (directories.target);
    disp ('Created.');
end;



% Install.
fprintf ([banner 'Install ' octave.out ' in ' directories.target ' ... ']);
movefile (octave.out, directories.target, 'f');
disp ('Done.');



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
